### ISSUE TITLE
Iteration of `Char`

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#11479]
When iterating of values of `Char` via `#succ`, `#pred` or `#step`, all possible values are treated like a consecutive sequence from `Char::ZERO` to `Char::MAX`. But this is actually not true: There's a gap in between. The high surrogate and low surrogate codepoints `U+D800` through `U+DFFF` are not valid `Char` values.

Currently, you get an error when iterating into (or through) this gap:
```cr
'\uD7FF'.succ # ArgumentError: 0xd800 out of char range
```

This should be fixed such that the gap between `U+D7FF` and `U+E000` is transparently skipped:
```cr
'\uD7FF'.succ # => '\uE000`
'\uD7FF'.step(to: '\uE000`).to_a # => ['\uD7FF', '\uE000`]
```

This means that `'\uD7FF'.succ` has different semantics than `'\uD7FF' + 1`. The latter must stay an error because arithmatic operations are defined on codepoints. But we can make sure to iterate only the valid ranges.

**Comment by HertzDevil** [comment]
Apart from ranges I think it makes sense that `"\uD7FF".succ` should be `"\uE000"`.

**Comment by straight-shoota** [comment]
Yeah, that totally slipped my mind. This applies to `String` as well.

**Comment by HertzDevil** [comment]
I have confirmed that this is also the behaviour in Ruby:

```ruby
def succ(str)
  str.succ.codepoints.map { |x| "U+%04X" % x }.join(' ')
end

succ("\u{D7FF}") # => "U+E000"
```

**Comment by HertzDevil** [comment]
If `Steppable` is defined in terms of `self + step` and `Char#+` continues to raise, how do we solve that? Could we introduce a step parameter to `#succ` itself instead?
